File: Cocoa/Plug-Ins/Sound Editor/SoundWindowController.m
#import "SoundWindowController.h"

@implementation SoundWindowController

- (instancetype)initWithResource:(id <ResKnifeResource>)newResource {
    self = [self initWithWindowNibName:@"SoundWindow"];
    if (!self) return nil;
    
    _resource = newResource;
    if (_resource.data.length) {
        self.sound = [[SoundResource alloc] initWithResource:_resource];
    }
    
    [self window];
    return self;
}

- (void)windowDidLoad {
    [super windowDidLoad];
    self.channels.stringValue = self.sampleRate.stringValue = self.duration.stringValue = @"";
    if (!self.sound) {
        self.format.stringValue = @"(empty)";
    } else if (self.sound.format) {
        NSString *format = GetNSStringFromOSType(self.sound.format);
        if (self.sound.valid) {
            self.format.stringValue = SoundResource.supportedFormats[format];
            self.duration.stringValue = [self stringFromSeconds:self.sound.duration];
            self.playButton.enabled = YES;
            self.exportButton.enabled = YES;
            [self.sound play];
        } else {
            self.format.stringValue = [NSString stringWithFormat:@"%@ (unsupported)", format];
        }
        self.channels.stringValue = self.sound.channels == 2 ? @"Stereo" : @"Mono";
        self.sampleRate.stringValue = [NSString stringWithFormat:@"%.0f Hz", self.sound.sampleRate];
    } else {
        self.format.stringValue = @"(unknown)";
    }
}

- (NSString *)stringFromSeconds:(double)seconds {
    if (seconds < 10) {
        return [NSString stringWithFormat:@"%0.2fs", seconds];
    }
    if (seconds < 60) {
        return [NSString stringWithFormat:@"%0.1fs", seconds];
    }
    long s = (long)round(seconds);
    return [NSString stringWithFormat:@"%02ld:%02ld", s/60, s%60];
}

- (void)dealloc {
    [self.sound stop];
}

- (IBAction)playSound:(id)sender {
    [self.sound play];
}

- (IBAction)exportSound:(id)sender {
    NSSavePanel *panel = [NSSavePanel savePanel];
    if (_resource.name.length) {
        panel.nameFieldStringValue = _resource.name;
    } else {
        panel.nameFieldStringValue = [NSString stringWithFormat:@"Sound %d", _resource.resID];
    }
    panel.allowedFileTypes = @[@"aiff"];
    [panel beginSheetModalForWindow:self.window completionHandler:^(NSModalResponse returnCode) {
        if (returnCode == NSFileHandlingPanelOKButton) {
            [self.sound exportToURL:panel.URL];
        }
    }];
}

- (IBAction)importSound:(id)sender {
    NSOpenPanel *panel = [NSOpenPanel openPanel];
    panel.allowedFileTypes = @[@"public.audio"];
    [panel beginSheetModalForWindow:self.window completionHandler:^(NSModalResponse returnCode) {
        if (returnCode == NSFileHandlingPanelOKButton) {
            self.sound = [[SoundResource alloc] initWithURL:panel.URL format:'twos' channels:0 sampleRate:22050];
        }
    }];
}

@end

File: Cocoa/Plug-Ins/Sound Editor/SoundWindowController.swift
import AVKit
import Cocoa

class SoundWindowController: NSWindowController, ResKnifePlugin {
    let resource: ResKnifeResource
    let sound: SoundResource
    @IBOutlet var playButton: NSButton!
    @IBOutlet var exportButton: NSButton!
    @IBOutlet var format: NSTextField!
    @IBOutlet var channels: NSTextField!
    @IBOutlet var sampleRate: NSTextField!
    @IBOutlet var duration: NSTextField!
    @IBOutlet var accessoryView: NSView!
    @IBOutlet var selectFormat: NSPopUpButton!
    @IBOutlet var selectChannels: NSPopUpButton!
    @IBOutlet var selectSampleRate: NSPopUpButton!

    override var windowNibName: String! {
        return "SoundWindow"
    }
    
    required init(resource: ResKnifeResource) {
        UserDefaults.standard.register(defaults: ["SndFormat":k16BitBigEndianFormat])
        self.resource = resource
        sound = SoundResource(resource.data!)
        super.init(window: nil)

        NotificationCenter.default.addObserver(self, selector: #selector(self.soundDidPlayStop), name: .SoundDidStartPlaying, object: sound)
        NotificationCenter.default.addObserver(self, selector: #selector(self.soundDidPlayStop), name: .SoundDidStopPlaying, object: sound)
        self.window?.makeKeyAndOrderFront(self)
    }
    
    required init?(coder: NSCoder) {
        fatalError("not implemented")
    }
    
    deinit {
        sound.stop()
    }

    override func windowDidLoad() {
        super.windowDidLoad()
        self.format.stringValue = ""
        self.channels.stringValue = ""
        self.sampleRate.stringValue = ""
        self.duration.stringValue = ""
        self.loadInfo()
        sound.play()
    }
    
    func loadInfo() {
        if resource.data!.count == 0 {
            self.format.stringValue = "(empty)"
        } else if sound.format != 0 {
            if sound.valid {
                self.format.stringValue = SoundResource.supportedFormats[sound.format]!
                self.duration.stringValue = stringFromSeconds(sound.duration)
                playButton.isEnabled = true
                exportButton.isEnabled = true
            } else {
                let format = NSFileTypeForHFSTypeCode(OSType(sound.format))!
                self.format.stringValue = "\(format) (unsupported)"
            }
            self.channels.stringValue = sound.channels == 2 ? "Stereo" : "Mono"
            self.sampleRate.stringValue = String(format: "%.0f Hz", sound.sampleRate)
        } else {
            self.format.stringValue = "(unknown)"
        }
    }
    
    func stringFromSeconds(_ seconds: Double) -> String {
        if seconds < 10 {
            return String(format: "%0.2fs", seconds)
        }
        if seconds < 60 {
            return String(format: "%0.1fs", seconds)
        }
        let s = Int(round(seconds))
        return String(format: "%02ld:%02ld", s / 60, s % 60)
    }
    
    @objc func soundDidPlayStop(_ notification: Notification) {
        DispatchQueue.main.async {
            self.playButton.title = self.sound.playing ? "Stop" : "Play"
        }
    }

    @IBAction func playSound(_ sender: Any) {
        sound.playing ? sound.stop() : sound.play()
    }
    
    @IBAction func exportSound(_ sender: Any) {
        let panel = NSSavePanel()
        if self.resource.name!.count > 0 {
            panel.nameFieldStringValue = self.resource.name!
        } else {
            panel.nameFieldStringValue = "Sound \(resource.resID)"
        }
        panel.allowedFileTypes = ["aiff"]
        panel.beginSheetModal(for: self.window!, completionHandler: { returnCode in
            if returnCode.rawValue == NSFileHandlingPanelOKButton {
                self.sound.export(to: panel.url!)
            }
        })
    }
    
    @IBAction func importSound(_ sender: Any) {
        let panel = NSOpenPanel()
        panel.allowedFileTypes = ["public.audio"]
        panel.accessoryView = self.accessoryView
        panel.isAccessoryViewDisclosed = true
        panel.prompt = "Import"
        panel.beginSheetModal(for: self.window!, completionHandler: { returnCode in
            if returnCode.rawValue == NSFileHandlingPanelOKButton {
                let format = self.selectFormat.selectedTag()
                let channels = self.selectChannels.selectedTag()
                let sampleRate = self.selectSampleRate.selectedTag()
                self.sound.load(url: panel.url!, format: UInt32(format), channels: UInt32(channels), sampleRate: Double(sampleRate))
                do {
                    self.resource.data = try self.sound.data()
                } catch {

                }
                self.setDocumentEdited(true)
                self.loadInfo()
            }
        })
    }
}
